include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,291 @@
+  {
+    RMM_ASSERT_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+
+    read_lock lock(mtx_);
+
+    if (is_default_stream(stream)) {
+      auto id = std::this_thread::get_id();
+      for (auto& kv : thread_arenas_) {
+        // Check the per-thread arena if it does not belong to the current thread, and return if the
+        // pointer is found.
+        if (kv.first != id && kv.second.deallocate(p, bytes)) return;
+      }
+    } else {
+      for (auto& kv : stream_arenas_) {
+        // Check the per-stream arena if it does not belong to the current stream, and return if the
+        // pointer is found.
+        if (kv.first != stream && kv.second.deallocate(p, bytes)) return;
+      }
+    }
+
+    // Allocation not found.
+    RMM_LOGGING_ASSERT(0);
+  }
+
+  /**
+   * @brief Get the arena associated with the current thread or the given stream.
+   *
+   * @param stream The stream associated with the arena.
+   * @return arena& The arena associated with the current thread or the given stream.
+   */
+  arena& get_arena(cudaStream_t stream)
+  {
+    if (is_default_stream(stream)) {
+      return get_thread_arena();
+    } else {
+      return get_stream_arena(stream);
+    }
+  }
+
+  /**
+   * @brief Get the arena associated with the current thread.
+   *
+   * @return arena& The arena associated with the current thread.
+   */
+  arena& get_thread_arena()
+  {
+    auto id = std::this_thread::get_id();
+    {
+      read_lock lock(mtx_);
+      auto it = thread_arenas_.find(id);
+      if (it != thread_arenas_.end()) { return it->second; }
+    }
+    {
+      write_lock lock(mtx_);
+      thread_arenas_.emplace(id, global_arena_);
+      return thread_arenas_.at(id);
+    }
+  }
+
+  /**
+   * @brief Get the arena associated with the given stream.
+   *
+   * @return arena& The arena associated with the given stream.
+   */
+  arena& get_stream_arena(cudaStream_t stream)
+  {
+    RMM_LOGGING_ASSERT(!is_default_stream(stream));
+    {
+      read_lock lock(mtx_);
+      auto it = stream_arenas_.find(stream);
+      if (it != stream_arenas_.end()) { return it->second; }
+    }
+    {
+      write_lock lock(mtx_);
+      stream_arenas_.emplace(stream, global_arena_);
+      return stream_arenas_.at(stream);
+    }
+  }
+
+  /**
+   * @brief Get free and available memory for memory resource.
+   *
+   * @param stream to execute on.
+   * @return std::pair containing free_size and total_size of memory.
+   */
+  std::pair<std::size_t, std::size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    return std::make_pair(0, 0);
+  }
+
+  /**
+   * @brief Check if the given stream is considered a "default" stream.
+   *
+   * @param stream to check.
+   * @return true if the given stream is a default stream, false otherwise.
+   */
+  static bool is_default_stream(cudaStream_t stream)
+  {
+    return stream == cudaStreamDefault || stream == cudaStreamLegacy ||
+           stream == cudaStreamPerThread;
```

Comment:
I'm uneasy about grouping `cudaStreamLegacy` in with `cudaStreamDefault` and `cudaStreamPerThread`.

`cudaStreamLegacy` must _always_ maintain the implicit synchronization behavior of the null stream. Is that being done in your implementation?

Reply:
This is really checking if we should use the per-thread arena for this stream. I've renamed the method to make it more clear.

Reply:
That's my point. `cudaStreamLegacy` should not use a per-thread arena. Unlike `cudaStreamLegacy` in PTDS mode or `cudaStreamPerThread`, there is only _one_ `cudaStreamLegacy`, not one per thread.

Reply:
OK, I think now it should match you are saying. Please take another look.